File: rocket-bi-web/di-core/lake-house/domain/lake-job/LakeJobHistory.ts
import { LakeJobStatus } from '@core/lake-house/domain/lake-job/LakeJobStatus';

export class LakeJobHistory {
  constructor(
    public runId: number,
    public jobId: number,
    public jobName: string,
    public startTime: number,
    public updatedTime: number,
    public endTime: number,
    public jobStatus: LakeJobStatus,
    public message: string,
    public logPath?: string
  ) {}

  get totalRuntime(): number {
    const totalRunTime = this.updatedTime - this.startTime;
    return totalRunTime > 0 ? totalRunTime : 0;
  }

  static fromObject(obj: any) {
    return new LakeJobHistory(obj.runId, obj.jobId, obj.jobName, obj.startTime, obj.updatedTime, obj.endTime, obj.jobStatus, obj.message, obj.logPath);
  }
}

File: job-scheduler/src/main/scala/datainsider/lakescheduler/domain/LakeJobHistory.scala
package datainsider.lakescheduler.domain

import com.fasterxml.jackson.annotation.JsonSubTypes.Type
import com.fasterxml.jackson.annotation.{JsonSubTypes, JsonTypeInfo}
import com.fasterxml.jackson.module.scala.JsonScalaEnumeration
import datainsider.jobscheduler.domain.Ids.{JobId, OrgId, RunId}
import datainsider.lakescheduler.domain.job.LakeJobStatus.LakeJobStatus
import datainsider.lakescheduler.domain.job.LakeJobStatusRef

case class LakeJobHistory(
    runId: RunId = 0,
    jobId: JobId,
    jobName: String,
    yarnAppId: String,
    startTime: Long,
    updatedTime: Long,
    endTime: Long,
    @JsonScalaEnumeration(classOf[LakeJobStatusRef]) jobStatus: LakeJobStatus,
    message: String
)

@JsonTypeInfo(
  use = JsonTypeInfo.Id.NAME,
  include = JsonTypeInfo.As.PROPERTY,
  property = "class_name"
)
@JsonSubTypes(
  Array(
    new Type(value = classOf[JavaJobProgress], name = "java_job_progress"),
    new Type(value = classOf[SqlJobProgress], name = "sql_job_progress")
  )
)
trait LakeJobProgress {

  def orgId: Long

  def runId: Long

  def jobId: JobId

  def startTime: Long

  def updatedTime: Long

  @JsonScalaEnumeration(classOf[LakeJobStatusRef])
  def jobStatus: LakeJobStatus

  def progressData: Map[String, Any]

  def message: Option[String]
}

case class JavaJobProgress(
    orgId: OrgId,
    runId: RunId,
    jobId: JobId,
    yarnAppId: Option[String],
    startTime: Long,
    updatedTime: Long,
    jobStatus: LakeJobStatus,
    message: Option[String] = None
) extends LakeJobProgress {
  override def progressData: Map[String, Any] = {
    Map("yarn_app_id" -> yarnAppId.getOrElse(""))
  }
}

case class SqlJobProgress(
    orgId: OrgId,
    runId: RunId,
    jobId: JobId,
    yarnAppId: Option[String],
    startTime: Long,
    updatedTime: Long,
    jobStatus: LakeJobStatus,
    message: Option[String] = None
) extends LakeJobProgress {
  override def progressData: Map[String, Any] = {
    Map("yarn_app_id" -> yarnAppId.getOrElse(""))
  }
}
